File: packages/core/platforms/lastlink.py
"""Adapter Lastlink.

Lastlink envia POST com schema (referência: docs Lastlink, formato 2024+):
{
  "Id": "...",
  "Event": "Purchase_Order_Confirmed" | "Purchase_Request_Confirmed" |
           "Purchase_Request_Expired" | "Subscription_Canceled" | ...,
  "CreatedAt": "...",
  "Data": {
    "Buyer": {"Name", "Email", "PhoneNumber", "Document"},
    "Products": [{"Id", "Name", ...}],
    "Payment": {
      "Method": "CreditCard" | "Pix" | "Bankslip",
      "PixCode" / "PixCopyPaste"?,
      "PixExpirationDate"?,
      "BankSlipUrl"?,
      "BankSlipBarcode"?,
      "BankSlipDueDate"?,
      "Amount" (em reais),
      ...
    },
    "Subscriptions": [{...}]?,
    "Offer": {"Url"?},
  }
}

Validação: header `X-Lastlink-Signature` (HMAC-SHA256 hex sobre raw body).
Se Lastlink tiver schema diferente na sua conta, ajuste os campos abaixo —
o teste smoke valida com payload sintético; rode primeiro contra um webhook real
e ajuste se necessário.
"""
import hashlib
import hmac
from typing import Any, Mapping

from .base import (
    Customer,
    EventKind,
    NormalizedEvent,
    Payment,
    Product,
    first_name_of,
    normalize_phone_br,
)


class LastlinkAdapter:
    name = "lastlink"

    def validate_signature(
        self, raw_body: bytes, headers: Mapping[str, str], query_params: Mapping[str, str], secret: str
    ) -> bool:
        if not secret:
            return True
        sig = headers.get("x-lastlink-signature") or headers.get("x-hub-signature-256", "")
        # Alguns setups mandam "sha256=<hex>" — extrai
        if sig.startswith("sha256="):
            sig = sig.split("=", 1)[1]
        expected = hmac.new(secret.encode(), raw_body, hashlib.sha256).hexdigest()
        return hmac.compare_digest(expected, sig)

    def event_kind(self, payload: dict[str, Any]) -> EventKind:
        evt = (payload.get("Event") or "").lower()
        method = (((payload.get("Data") or {}).get("Payment") or {}).get("Method") or "").lower()

        if evt in ("purchase_order_confirmed", "purchase_approved"):
            return "compra_aprovada"
        if evt in ("purchase_request_confirmed", "purchase_pending_payment", "purchase_request_created"):
            if "pix" in method:
                return "pix"
            if "bankslip" in method or "boleto" in method:
                return "boleto"
            return "outro"
        if evt in ("purchase_request_expired", "purchase_canceled", "subscription_canceled"):
            return "cancelada"
        if evt in ("purchase_refused", "purchase_failed"):
            return "recusada"
        if evt in ("abandoned_cart", "purchase_abandoned"):
            return "carrinho"
        if evt in ("subscription_renewed",):
            return "renovacao"
        return "outro"

    def normalize(self, payload: dict[str, Any]) -> NormalizedEvent:
        data = payload.get("Data") or {}
        buyer = data.get("Buyer") or {}
        products = data.get("Products") or []
        first_product = products[0] if products else {}
        payment = data.get("Payment") or {}
        offer = data.get("Offer") or {}

        full_name = buyer.get("Name") or ""
        phone = normalize_phone_br(buyer.get("PhoneNumber") or buyer.get("Phone") or "")
        try:
            value_cents = int(round(float(payment.get("Amount") or payment.get("TotalAmount") or 0) * 100))
        except (TypeError, ValueError):
            value_cents = 0

        return NormalizedEvent(
            platform=self.name,
            event_kind=self.event_kind(payload),
            raw_event_type=payload.get("Event", ""),
            customer=Customer(
                name=full_name,
                first_name=first_name_of(full_name),
                email=buyer.get("Email", ""),
                phone=phone,
                user_number=f"{phone}@s.whatsapp.net" if phone else "",
            ),
            product=Product(
                name=first_product.get("Name", ""),
                id=str(first_product.get("Id") or ""),
                value_cents=value_cents,
            ),
            payment=Payment(
                pix_code=payment.get("PixCode") or payment.get("PixCopyPaste") or "",
                pix_expiration=payment.get("PixExpirationDate") or "",
                boleto_url=payment.get("BankSlipUrl") or payment.get("BoletoUrl") or "",
                boleto_barcode=payment.get("BankSlipBarcode") or payment.get("BoletoBarcode") or "",
                boleto_expiry=payment.get("BankSlipDueDate") or payment.get("BoletoExpiry") or "",
                access_url=offer.get("Url") or data.get("AccessUrl") or "",
                rejection_reason=payment.get("RefusalReason") or "",
                method=(payment.get("Method") or "").lower(),
            ),
            raw_payload=payload,
        )

File: worker/src/adapters/lastlink.ts
import { hmacSha256Hex, safeEqual } from "../crypto";
import type { Address, EventKind, NormalizedEvent, PlatformAdapter, Tracking } from "../types";
import { firstNameOf, normalizePhoneBr } from "../types";

export const LastlinkAdapter: PlatformAdapter = {
  name: "lastlink",

  async validateSignature(rawBody, headers, _query, secret) {
    if (!secret) return true;
    let sig =
      headers.get("x-lastlink-signature") ||
      headers.get("x-lastlink-token") ||
      headers.get("x-hub-signature-256") ||
      "";
    if (sig.startsWith("sha256=")) sig = sig.slice(7);
    if (!sig) return false;
    const expected = await hmacSha256Hex(secret, rawBody);
    return safeEqual(expected, sig);
  },

  normalize(payload) {
    const data = payload.Data || {};
    const buyer = data.Buyer || {};
    const products: any[] = data.Products || [];
    const firstProduct = products[0] || {};
    const purchase = data.Purchase || {};
    const purchasePayment = purchase.Payment || {};
    const price = purchase.Price || purchase.OriginalPrice || {};
    const pix = purchase.Pix || {};
    const bankSlip = purchase.BankSlip || purchase.Boleto || {};
    const offer = data.Offer || {};
    const subscriptions: any[] = data.Subscriptions || [];
    const subscription = subscriptions[0] || {};
    const utm = data.Utm || {};
    const affiliate = purchase.Affiliate || {};
    const buyerAddr = buyer.Address || {};

    const fullName = buyer.Name || "";
    const phone = normalizePhoneBr(buyer.PhoneNumber || buyer.Phone || "");
    const valueCents = Math.round(parseFloat(price.Value || "0") * 100) || 0;
    const method = (purchasePayment.PaymentMethod || purchasePayment.Method || "").toLowerCase();

    const address: Address | undefined = buyerAddr.Street || buyerAddr.City
      ? {
          zipCode: buyerAddr.ZipCode || "",
          street: buyerAddr.Street || "",
          streetNumber: buyerAddr.StreetNumber || "",
          complement: buyerAddr.Complement || "",
          district: buyerAddr.District || "",
          city: buyerAddr.City || "",
          state: buyerAddr.State || "",
        }
      : undefined;

    const tracking: Tracking | undefined = utm.UtmSource || utm.UtmMedium || affiliate.Id
      ? {
          utmSource: utm.UtmSource || "",
          utmMedium: utm.UtmMedium || "",
          utmCampaign: utm.UtmCampaign || "",
          utmTerm: utm.UtmTerm || "",
          utmContent: utm.UtmContent || "",
          src: utm.Src || "",
          sck: utm.Sck || "",
          vtid: utm.Vtid || "",
          affiliateId: affiliate.Id || "",
          affiliateEmail: affiliate.Email || "",
        }
      : undefined;

    return {
      platform: "lastlink",
      eventKind: mapKind(payload.Event || "", method),
      rawEventType: payload.Event || "",
      isTest: Boolean(payload.IsTest),
      customer: {
        name: fullName,
        firstName: firstNameOf(fullName),
        email: buyer.Email || "",
        phone,
        document: buyer.Document || "",
        address,
      },
      product: {
        name: firstProduct.Name || "",
        id: String(firstProduct.Id || ""),
        valueCents,
      },
      payment: {
        pixCode: pix.QrCodeText || "",
        pixQrUrl: pix.QrCode || "",
        pixExpiration: pix.ExpirationDate || pix.ExpiresAt || "",
        // Lastlink BankSlip schema:
        // - DigitableLine = linha digitável (humano)
        // - BarCodeData = código de barras numérico puro
        // - BarCode = URL da imagem do código de barras
        // Não há "URL do boleto" — usar InvoiceUrl como fallback
        boletoUrl: purchase.InvoiceUrl || bankSlip.Url || bankSlip.BankSlipUrl || "",
        boletoBarcode: bankSlip.DigitableLine || bankSlip.Barcode || bankSlip.BankSlipBarcode || "",
        boletoBarcodeRaw: bankSlip.BarCodeData || "",
        boletoBarcodeImage: bankSlip.BarCode || "",
        boletoExpiry: bankSlip.DueDate || bankSlip.BankSlipDueDate || subscription.NextBilling || "",
        accessUrl: offer.Url || subscription.SubscriberUrl || "",
        invoiceUrl: purchase.InvoiceUrl || "",
        rejectionReason: purchasePayment.RefusalReason || "",
        method,
        installments: purchasePayment.NumberOfInstallments || undefined,
        paymentId: purchase.PaymentId || "",
        paymentDate: purchase.PaymentDate || "",
      },
      tracking,
      subscription:
        subscription.Id ||
        purchase.NextBilling ||
        purchase.ChargebackDate ||
        subscription.CanceledDate ||
        subscription.ExpiredDate
          ? {
            id: subscription.Id || "",
            recurrency: typeof purchase.Recurrency === "number" ? purchase.Recurrency : undefined,
            status: subscription.Status || "",
            nextBilling: purchase.NextBilling || "",
            canceledDate: subscription.CanceledDate || "",
            expiredDate: subscription.ExpiredDate || "",
            cancellationReason: subscription.CancellationReason || "",
            chargebackDate: purchase.ChargebackDate || "",
          }
          : undefined,
      rawPayload: payload,
    };
  },
};

function mapKind(rawEvt: string, rawMethod: string): EventKind {
  const e = rawEvt.toLowerCase();
  const m = rawMethod.toLowerCase();

  if (
    e === "purchase_order_confirmed" ||
    e === "purchase_order_completed" ||
    e === "purchase_approved" ||
    e === "purchase_complete" ||
    e === "membership_started" ||
    e === "access_release_started"
  ) {
    return "compra_aprovada";
  }

  if (
    e === "purchase_request_confirmed" ||
    e === "purchase_request_created" ||
    e === "purchase_pending_payment" ||
    e === "invoice_created"
  ) {
    if (m.includes("pix")) return "pix";
    if (m.includes("bankslip") || m.includes("boleto")) return "boleto";
    return "outro";
  }

  if (e === "subscription_renewal_pending") {
    if (m.includes("pix")) return "pix";
    if (m.includes("bankslip") || m.includes("boleto")) return "boleto";
    return "outro";
  }

  if (
    e === "subscription_renewal_approved" ||
    e === "subscription_renewed" ||
    e === "renewal_payment_completed"
  ) {
    return "renovacao";
  }

  if (
    e === "purchase_request_expired" ||
    e === "purchase_expired" ||
    e === "purchase_order_expired"
  ) {
    if (m.includes("pix")) return "pix";
    if (m.includes("bankslip") || m.includes("boleto")) return "boleto";
    return "cancelada";
  }

  if (
    e === "purchase_canceled" ||
    e === "purchase_cancelled" ||
    e === "purchase_request_canceled" ||
    e === "purchase_request_cancelled" ||
    e === "subscription_canceled" ||
    e === "subscription_expired" ||
    e === "payment_chargeback" ||
    e === "payment_refunded" ||
    e === "payment_refund" ||
    e === "payment_reversal" ||
    e === "payment_reversed" ||
    e === "membership_ended" ||
    e === "access_release_ended" ||
    e === "refund_period_ended" ||
    e === "refund_requested"
  ) {
    return "cancelada";
  }

  if (e === "purchase_refused" || e === "purchase_failed" || e === "payment_failed") {
    return "recusada";
  }

  if (e === "abandoned_cart" || e === "purchase_abandoned" || e === "cart_abandoned") {
    return "carrinho";
  }

  // Refund_Period_Over: cliente passou prazo de 7 dias, "venda firmada"
  // Não é cancelada (cliente continua), não é recompra. Marca como "outro" — workflow custom via raw_event_type
  if (e === "refund_period_over" || e === "refund_period_ended") {
    return "outro";
  }

  return "outro";
}
